arangod/Pregel/Algos/AIR/Greenspun/Primitives.cpp

```diff
@@ -0,0 +1,1264 @@
+    }
+
+    result.add(VPackArrayIterator(list));
+    for (iter.next(); iter.valid(); iter.next()) {
+      result.add(*iter);
+    }
+  }
+
+  return {};
+}
+
+EvalResult Prim_Assert(Machine& ctx, VPackSlice const paramsList, VPackBuilder& result) {
+  VPackArrayIterator iter(paramsList);
+  if (!iter.valid()) {
+    return EvalError("expected at least one argument");
+  }
+
+  VPackSlice value = *iter;
+  if (ValueConsideredFalse(value)) {
+    iter++;
+    std::string errorMessage = "assertion failed";
+    if (iter.valid()) {
+      errorMessage = paramsToString(iter);
+    }
+    return EvalError(errorMessage);
+  }
+
+  result.add(VPackSlice::noneSlice());
+  return {};
+}
+
+EvalResult Prim_Sort(Machine& ctx, VPackSlice const paramsList, VPackBuilder& result) {
+  auto res = extract<VPackSlice, VPackSlice>(paramsList);
+  if (!res) {
+    return std::move(res).asResult();
+  }
+
+  auto&& [func, list] = res.value();
+  if (!list.isArray()) {
+    return EvalError("expected list as second parameter, found: " + list.toJson());
+  }
+
+  VPackArrayIterator iter(list);
+  std::vector<VPackSlice> v;
+  v.reserve(list.length());
+  v.assign(iter.begin(), iter.end());
+
+  struct Compare {
+    Machine& ctx;
+    VPackSlice func;
+
+    // return true if A is _less_ than B
+    bool operator()(VPackSlice A, VPackSlice B) {
+      VPackBuilder parameter;
+      {
+        VPackArrayBuilder pb(&parameter);
+        parameter.add(A);
+        parameter.add(B);
+      }
+
+      VPackBuilder tempBuffer;
+
+      auto res = EvaluateApply(ctx, func, VPackArrayIterator(parameter.slice()),
+                               tempBuffer, false);
+      if (res.fail()) {
+        throw res.error().wrapMessage("when mapping pair " + parameter.toJson());
+      }
+
+      return ValueConsideredTrue(tempBuffer.slice());
+    }
+  };
+
+  try {
+    std::sort(v.begin(), v.end(), Compare{ctx, func});
+  } catch (EvalError& err) {
+    return err.wrapMessage("in compare function");
+  }
+
+  VPackArrayBuilder ab(&result);
+  for (auto&& slice : v) {
+    result.add(slice);
+  }
+  return {};
+}
+
+double rand_source_query() {
+  return static_cast<double>(std::rand()) / RAND_MAX;
+}
+
+EvalResult Prim_Rand(Machine& ctx, VPackSlice const paramsList, VPackBuilder& result) {
+  result.add(VPackValue(rand_source_query()));
+  return {};
+}
+
+EvalResult Prim_RandRange(Machine& ctx, VPackSlice const paramsList, VPackBuilder& result) {
+  auto res = extract<double, double>(paramsList);
+  if (!res) {
+    return res.error();
+  }
+
+  auto&& [min, max] = res.value();
+
+  auto v = min + rand_source_query() * (max - min);
+  result.add(VPackValue(v));
+  return {};
+}
+
+void RegisterFunction(Machine& ctx, std::string_view name, Machine::function_type&& f) {
+  ctx.setFunction(name, std::move(f));
+}
+
+void RegisterAllPrimitives(Machine& ctx) {
+  // Calculation operators
+  ctx.setFunction("banana", Prim_Banana);
```

Comment:
Why is this named banana and not "sum"?
```suggestion
  ctx.setFunction("sum", Prim_Sum);
```